File: electron/ipc/videoGenerate.ts
import { ipcMain } from 'electron'
import { join, dirname } from 'path'
import { existsSync } from 'fs'
import { promises as fs } from 'fs'
import { exec } from 'child_process'
import { execa } from 'execa'
import type { GenerateImageParams } from '../../shared/types.js'
import type { AppState } from './state.js'
import { loadModelGroups } from './modelGroups.js'
import { AsyncOperationGuard } from '../utils/AsyncOperationGuard.js'
import { ResourceManager } from '../utils/ResourceManager.js'
import { getDefaultOutputsFolder, getDefaultSDCppFolder, getFFmpegPath, resolveModelPath } from '../utils/paths.js'

const VIDEO_DEFAULT_NEGATIVE_PROMPT =
  '色调艳丽，过曝，静态，细节模糊不清，字幕，风格，作品，画作，画面，静止，整体发灰，最差质量，低质量，JPEG压缩残留，丑陋的，残缺的，多余的手指，画得不好的手部，画得不好的脸部，畸形的，毁容的，形态畸形的肢体，手指融合，静止不动的画面，杂乱的背景，三条腿，背景人很多，倒着走'

interface VideoGenerateDeps {
  state: AppState
}

function getSDCppExecutablePath(state: AppState): string {
  const engineFolder = state.sdcppFolderPath || getDefaultSDCppFolder()
  const deviceFolder = join(engineFolder, state.sdcppDeviceType)
  return join(deviceFolder, 'sd-cli.exe')
}

export function registerVideoGenerateHandlers({ state }: VideoGenerateDeps): void {
  // 取消视频生成
  ipcMain.handle('generate-video:cancel', async () => {
    if (state.currentGenerateProcess && !state.currentGenerateProcess.killed) {
      if (state.currentGenerateKill) state.currentGenerateKill()
      return true
    }
    return false
  })

  // 开始视频生成
  ipcMain.handle(
    'generate-video:start',
    async (event, params: GenerateImageParams & { frames?: number; fps?: number }) => {
      try {
        const {
          groupId,
          deviceType,
          mode,
          initImage,
          prompt,
          negativePrompt = '',
          steps = 20,
          width = 512,
          height = 512,
          cfgScale = 7.0,
          samplingMethod,
          scheduler,
          seed,
          batchCount = 1,
          threads,
          preview,
          previewInterval = 1,
          verbose = false,
          color = false,
          offloadToCpu = false,
          diffusionFa = false,
          controlNetCpu = false,
          clipOnCpu = false,
          vaeOnCpu = false,
          diffusionConvDirect = false,
          vaeConvDirect = false,
          vaeTiling = false,
          frames = 33,
          fps = 8,
          flowShift = 3.0,
          highNoiseSteps,
          highNoiseCfgScale,
          highNoiseSamplingMethod,
        } = params

        let sdModelPath: string | undefined
        let highNoiseModelPath: string | undefined
        let clipVisionModelPath: string | undefined

        if (groupId) {
          const groups = await loadModelGroups(state.weightsFolderPath)
          const group = groups.find((g) => g.id === groupId)
          if (!group || !group.sdModel) throw new Error('模型组配置错误')
          sdModelPath = resolveModelPath(group.sdModel, state.weightsFolderPath)
          if (!sdModelPath || !existsSync(sdModelPath)) throw new Error(`模型文件不存在: ${sdModelPath}`)

          if (group.highNoiseSdModel) highNoiseModelPath = resolveModelPath(group.highNoiseSdModel, state.weightsFolderPath)
          if (group.clipVisionModel) clipVisionModelPath = resolveModelPath(group.clipVisionModel, state.weightsFolderPath)
        } else {
          throw new Error('必须提供模型组ID')
        }

        const sdExePath = getSDCppExecutablePath(state)
        if (!existsSync(sdExePath)) throw new Error(`引擎文件不存在: ${sdExePath}`)

        const outputsDir = getDefaultOutputsFolder()
        if (!existsSync(outputsDir)) await fs.mkdir(outputsDir, { recursive: true })

        const timestamp = Date.now()
        const outputAviPath = join(outputsDir, `video_${timestamp}.avi`)
        const outputMp4Path = join(outputsDir, `video_${timestamp}.mp4`)
        const outputMetadataPath = join(outputsDir, `video_${timestamp}.json`)

        const args: string[] = ['-M', 'vid_gen', '--diffusion-model', sdModelPath, '--prompt', prompt]
        if (mode === 'image2video' && initImage) args.push('-i', initImage)

        if (groupId) {
          const groups = await loadModelGroups(state.weightsFolderPath)
          const group = groups.find((g) => g.id === groupId)
          if (group?.vaeModel) {
            const p = resolveModelPath(group.vaeModel, state.weightsFolderPath)
            if (p && existsSync(p)) args.push('--vae', p)
          }
          if (group?.llmModel) {
            const p = resolveModelPath(group.llmModel, state.weightsFolderPath)
            if (p && existsSync(p)) args.push('--t5xxl', p)
          }
        }

        args.push('--negative-prompt', negativePrompt || VIDEO_DEFAULT_NEGATIVE_PROMPT)
        args.push('--output', outputAviPath)
        if (highNoiseModelPath) args.push('--high-noise-diffusion-model', highNoiseModelPath)
        if (clipVisionModelPath) args.push('--clip_vision', clipVisionModelPath)
        if (frames > 0) args.push('--video-frames', frames.toString())
        if (flowShift !== undefined) args.push('--flow-shift', flowShift.toString())
        if (steps !== 20) args.push('--steps', steps.toString())
        if (highNoiseModelPath) args.push('--high-noise-steps', (highNoiseSteps || steps).toString())
        if (width !== 512) args.push('--width', width.toString())
        if (height !== 512) args.push('--height', height.toString())
        if (Math.abs(cfgScale - 7.0) > 0.0001) args.push('--cfg-scale', cfgScale.toString())
        if (highNoiseModelPath) args.push('--high-noise-cfg-scale', (highNoiseCfgScale || cfgScale).toString())
        if (samplingMethod?.trim()) args.push('--sampling-method', samplingMethod.trim())
        if (highNoiseModelPath)
          args.push('--high-noise-sampling-method', highNoiseSamplingMethod || samplingMethod || 'euler')
        if (scheduler?.trim()) args.push('--scheduler', scheduler.trim())
        if (seed !== undefined && seed >= 0) args.push('--seed', seed.toString())
        if (batchCount > 1) args.push('--batch-count', batchCount.toString())
        if (threads !== undefined && threads > 0) args.push('--threads', threads.toString())

        if (preview && preview !== 'none' && preview.trim() !== '') {
          const p = join(outputsDir, `preview_video_${timestamp}.png`)
          args.push('--preview', preview.trim())
          args.push('--preview-path', join(outputsDir, `preview_video_${timestamp}.png`))
          if (previewInterval > 1) args.push('--preview-interval', previewInterval.toString())
        }

        if (verbose) args.push('--verbose')
        if (color) args.push('--color')
        if (offloadToCpu) args.push('--offload-to-cpu')
        if (diffusionFa) args.push('--diffusion-fa')
        if (controlNetCpu) args.push('--control-net-cpu')
        if (clipOnCpu) args.push('--clip-on-cpu')
        if (vaeOnCpu) args.push('--vae-on-cpu')
        if (diffusionConvDirect) args.push('--diffusion-conv-direct')
        if (vaeConvDirect) args.push('--vae-conv-direct')
        if (vaeTiling) args.push('--vae-tiling')

        console.log(`[Generate Video] Starting video generation (${deviceType}): ${sdExePath}`)
        const startTime = Date.now()
        event.sender.send('generate-video:progress', { progress: '正在启动 SD.cpp 引擎...' })

        return new Promise((resolvePromise, reject) => {
          let stdout = ''
          let stderr = ''
          let isResolved = false
          const operationGuard = new AsyncOperationGuard()
          const resourceManager = new ResourceManager()
          let childProcess: any = null

          const killProcess = () => {
            if (isResolved) return
            if (childProcess && childProcess.pid) {
              if (process.platform === 'win32') exec(`taskkill /F /T /PID ${childProcess.pid}`)
              else childProcess.kill()
            }
          }

          state.currentGenerateKill = killProcess

          const cleanup = () => {
            operationGuard.invalidate()
            resourceManager.cleanupAll()
            if (state.currentGenerateProcess === childProcess) {
              state.currentGenerateProcess = null
              state.currentGenerateKill = null
            }
          }

          const handleVideoCompletion = async () => {
            try {
              if (!existsSync(outputAviPath)) throw new Error('未找到输出视频文件')
              event.sender.send('generate-video:progress', { progress: '正在转换视频格式 (AVI -> MP4)...' })
              const ffmpegPath = getFFmpegPath()
              if (existsSync(ffmpegPath)) {
                await execa(ffmpegPath, [
                  '-i',
                  outputAviPath,
                  '-c:v',
                  'libx264',
                  '-pix_fmt',
                  'yuv420p',
                  '-y',
                  outputMp4Path,
                ])
                await fs.unlink(outputAviPath).catch(() => {})
              } else {
                throw new Error(`未找到 FFmpeg 引擎: ${ffmpegPath}`)
              }

              const duration = Date.now() - startTime
              const metadata = {
                prompt,
                negativePrompt,
                steps,
                width,
                height,
                cfgScale,
                deviceType,
                groupId,
                timestamp,
                generatedAt: new Date().toISOString(),
                samplingMethod,
                scheduler,
                seed,
                batchCount,
                threads,
                frames,
                fps,
                commandLine: args.join(' '),
                duration,
                type: 'video',
                mediaType: 'video',
              }
              await fs.writeFile(outputMetadataPath, JSON.stringify(metadata, null, 2), 'utf-8')
              const videoUrl = `media:///${outputMp4Path.replace(/\\/g, '/')}`
              event.sender.send('generate-video:progress', { progress: `生成完成`, video: videoUrl })
              resolvePromise({ success: true, video: videoUrl, videoPath: outputMp4Path, duration })
            } catch (e: any) {
              reject(e)
            }
          }

          childProcess = execa(sdExePath, args, { cwd: dirname(sdExePath) })
          state.currentGenerateProcess = childProcess

          childProcess.stdout?.on('data', (data: Buffer) => {
            const text = data.toString('utf8')
            stdout += text
            if (operationGuard.check() && !event.sender.isDestroyed()) {
              event.sender.send('generate-video:cli-output', { type: 'stdout', text })
              const m = text.match(/progress[:\s]+(\d+)%/i)
              if (m) event.sender.send('generate-video:progress', { progress: `生成中... ${m[1]}%` })
            }
          })

          childProcess.stderr?.on('data', (data: Buffer) => {
            const text = data.toString('utf8')
            stderr += text
            if (operationGuard.check() && !event.sender.isDestroyed()) {
              event.sender.send('generate-video:cli-output', { type: 'stderr', text })
            }
          })

          childProcess.on('close', (code: number) => {
            cleanup()
            if (isResolved) return
            isResolved = true
            if (code === 0) handleVideoCompletion()
            else reject(new Error(`SD.cpp 退出，错误代码: ${code}`))
          })

          childProcess.on('error', (err: Error) => {
            cleanup()
            if (isResolved) return
            isResolved = true
            reject(err)
          })
        })
      } catch (error: any) {
        console.error('[Generate Video] Error:', error)
        throw error
      }
    },
  )
}

File: src-tauri/src/commands/video_generate.rs
use crate::state::{self, AppState};
use std::path::Path;
use std::process::Stdio;
use tauri::{AppHandle, Emitter, State};
use tokio::io::{AsyncBufReadExt, BufReader};

fn resolve_sdcpp_executable(device_folder: &Path) -> Option<std::path::PathBuf> {
    let candidates: &[&str] = if cfg!(target_os = "windows") {
        &["sd.exe", "sd-cli.exe", "sd_server.exe", "sd-server.exe"]
    } else {
        &["sd", "sd-cli", "sd_server", "sd-server"]
    };

    candidates
        .iter()
        .map(|name| device_folder.join(name))
        .find(|path| path.exists())
}

/// Start video generation
#[tauri::command]
pub async fn generate_video_start(
    value: serde_json::Value,
    app: AppHandle,
    state: State<'_, AppState>,
) -> Result<serde_json::Value, String> {
    let device_type = value["deviceType"].as_str().unwrap_or("cpu");
    let prompt = value["prompt"].as_str().unwrap_or("");
    let mode = value["mode"].as_str().unwrap_or("text2video");

    // Get SD.cpp executable path
    let sdcpp_folder = state
        .sdcpp_folder
        .lock()
        .unwrap()
        .clone()
        .ok_or("SD.cpp folder not set")?;

    let device_folder = Path::new(&sdcpp_folder).join(device_type);
    let exe_path = resolve_sdcpp_executable(&device_folder).ok_or_else(|| {
        format!(
            "SD.cpp executable not found in: {} (expected one of: sd/sd-cli/sd_server/sd-server)",
            device_folder.display()
        )
    })?;

    // Get output path
    let outputs_folder = state
        .outputs_folder
        .lock()
        .unwrap()
        .clone()
        .unwrap_or_else(|| {
            state::get_default_outputs_folder()
                .to_string_lossy()
                .to_string()
        });

    std::fs::create_dir_all(&outputs_folder).map_err(|e| e.to_string())?;

    let timestamp = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis();

    // SD.cpp outputs video as an avi file initially
    let output_avi_path = Path::new(&outputs_folder).join(format!("generated_{}.avi", timestamp));
    let output_mp4_path = Path::new(&outputs_folder).join(format!("generated_{}.mp4", timestamp));

    // Build CLI arguments for video
    let args = build_video_args(&value, &state, &output_avi_path)?;

    // Setup cancellation
    let (cancel_tx, mut cancel_rx) = tokio::sync::watch::channel(false);
    *state.video_generate_cancel.lock().unwrap() = Some(cancel_tx);

    // Spawn process
    let mut child = tokio::process::Command::new(&exe_path)
        .args(&args)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .kill_on_drop(true)
        .spawn()
        .map_err(|e| format!("Failed to start process: {}", e))?;

    let pid = child.id();

    // Read stdout
    let stdout = child.stdout.take();
    let app_stdout = app.clone();
    let stdout_task = tokio::spawn(async move {
        if let Some(stdout) = stdout {
            let reader = BufReader::new(stdout);
            let mut lines = reader.lines();
            while let Ok(Some(line)) = lines.next_line().await {
                if line.contains('%') || line.contains("step") || line.contains("Step") {
                    let _ = app_stdout.emit(
                        "generate-video:progress",
                        serde_json::json!({ "progress": line }),
                    );
                }
                let _ = app_stdout.emit(
                    "generate-video:cli-output",
                    serde_json::json!({ "type": "stdout", "text": line }),
                );
            }
        }
    });

    // Read stderr
    let stderr = child.stderr.take();
    let app_stderr = app.clone();
    let stderr_task = tokio::spawn(async move {
        if let Some(stderr) = stderr {
            let reader = BufReader::new(stderr);
            let mut lines = reader.lines();
            while let Ok(Some(line)) = lines.next_line().await {
                if line.contains('%') || line.contains("step") || line.contains("Step") {
                    let _ = app_stderr.emit(
                        "generate-video:progress",
                        serde_json::json!({ "progress": line }),
                    );
                }
                let _ = app_stderr.emit(
                    "generate-video:cli-output",
                    serde_json::json!({ "type": "stderr", "text": line }),
                );
            }
        }
    });

    let start = std::time::Instant::now();

    // Wait for process
    let exit_status = tokio::select! {
        status = child.wait() => status.map_err(|e| e.to_string())?,
        _ = cancel_rx.changed() => {
            let _ = child.kill().await;
            if cfg!(target_os = "windows") {
                if let Some(pid) = pid {
                    let _ = tokio::process::Command::new("taskkill")
                        .args(&["/F", "/T", "/PID", &pid.to_string()])
                        .output()
                        .await;
                }
            }
            return Ok(serde_json::json!({ "success": false, "error": "cancelled" }));
        }
    };

    let _ = stdout_task.await;
    let _ = stderr_task.await;

    if !exit_status.success() {
        return Ok(serde_json::json!({
            "success": false,
            "error": format!("Process exited with code: {:?}", exit_status.code())
        }));
    }

    // Convert AVI to MP4 using FFmpeg
    let ffmpeg_path = state::get_ffmpeg_path();
    let mut final_video_path = output_avi_path.clone();

    if output_avi_path.exists() && ffmpeg_path.exists() {
        let ffmpeg_result = tokio::process::Command::new(&ffmpeg_path)
            .args(&[
                "-i",
                &output_avi_path.to_string_lossy(),
                "-c:v",
                "libx264",
                "-preset",
                "fast",
                "-crf",
                "23",
                "-y",
                &output_mp4_path.to_string_lossy(),
            ])
            .output()
            .await;

        if let Ok(output) = ffmpeg_result {
            if output.status.success() && output_mp4_path.exists() {
                let _ = tokio::fs::remove_file(&output_avi_path).await;
                final_video_path = output_mp4_path.clone();
            }
        }
    }

    let duration = start.elapsed().as_millis() as u64;

    // Save metadata
    let metadata_path = final_video_path.with_extension("json");
    let metadata = serde_json::json!({
        "prompt": prompt,
        "negativePrompt": value.get("negativePrompt"),
        "steps": value.get("steps"),
        "cfgScale": value.get("cfgScale"),
        "width": value.get("width"),
        "height": value.get("height"),
        "seed": value.get("seed"),
        "mode": mode,
        "deviceType": device_type,
        "groupId": value.get("groupId"),
        "type": "video",
        "mediaType": "video",
        "duration": duration,
        "generatedAt": chrono::Utc::now().to_rfc3339(),
    });
    let _ = tokio::fs::write(
        &metadata_path,
        serde_json::to_string_pretty(&metadata).unwrap_or_default(),
    )
    .await;

    if final_video_path.exists() {
        let video_url = format!(
            "media://{}",
            urlencoding::encode(&final_video_path.to_string_lossy())
        );

        Ok(serde_json::json!({
            "success": true,
            "video": video_url,
            "videoPath": final_video_path.to_string_lossy(),
            "duration": duration
        }))
    } else {
        Ok(serde_json::json!({
            "success": false,
            "error": "Output video not found after generation"
        }))
    }
}

/// Cancel video generation
#[tauri::command]
pub async fn generate_video_cancel(state: State<'_, AppState>) -> Result<bool, String> {
    if let Some(cancel) = state.video_generate_cancel.lock().unwrap().take() {
        let _ = cancel.send(true);
    }
    Ok(true)
}

fn build_video_args(
    params: &serde_json::Value,
    state: &State<'_, AppState>,
    output_path: &Path,
) -> Result<Vec<String>, String> {
    let mut args = Vec::new();

    let weights_folder = state
        .weights_folder
        .lock()
        .unwrap()
        .clone()
        .unwrap_or_default();

    // Model paths from group
    if let Some(group_id) = params["groupId"].as_str() {
        let groups_path = state::get_model_groups_path(Some(&weights_folder));
        if groups_path.exists() {
            let data = std::fs::read_to_string(&groups_path).map_err(|e| e.to_string())?;
            let groups: Vec<serde_json::Value> =
                serde_json::from_str(&data).map_err(|e| e.to_string())?;
            if let Some(group) = groups.iter().find(|g| g["id"].as_str() == Some(group_id)) {
                if let Some(sd_model) = group["sdModel"].as_str() {
                    let resolved = state::resolve_model_path(sd_model, &weights_folder);
                    args.push("-m".to_string());
                    args.push(resolved);
                }
                if let Some(vae) = group["vaeModel"].as_str() {
                    if !vae.is_empty() {
                        args.push("--vae".to_string());
                        args.push(state::resolve_model_path(vae, &weights_folder));
                    }
                }
                if let Some(llm) = group["llmModel"].as_str() {
                    if !llm.is_empty() {
                        args.push("--clip_l".to_string());
                        args.push(state::resolve_model_path(llm, &weights_folder));
                    }
                }
                if let Some(clip_vision) = group["clipVisionModel"].as_str() {
                    if !clip_vision.is_empty() {
                        args.push("--clip-vision".to_string());
                        args.push(state::resolve_model_path(clip_vision, &weights_folder));
                    }
                }
                if let Some(t5xxl) = group["t5xxlModel"].as_str() {
                    if !t5xxl.is_empty() {
                        args.push("--t5xxl".to_string());
                        args.push(state::resolve_model_path(t5xxl, &weights_folder));
                    }
                }
                if let Some(high_noise) = group["highNoiseSdModel"].as_str() {
                    if !high_noise.is_empty() {
                        args.push("--high-noise-model".to_string());
                        args.push(state::resolve_model_path(high_noise, &weights_folder));
                    }
                }
            }
        }
    }

    // Mode
    let mode = params["mode"].as_str().unwrap_or("text2video");
    args.push("--mode".to_string());
    args.push(mode.to_string());

    // Prompt
    if let Some(prompt) = params["prompt"].as_str() {
        if !prompt.is_empty() {
            args.push("-p".to_string());
            args.push(prompt.to_string());
        }
    }

    // Negative prompt
    if let Some(neg) = params["negativePrompt"].as_str() {
        if !neg.is_empty() {
            args.push("-n".to_string());
            args.push(neg.to_string());
        }
    }

    // Dimensions
    if let Some(w) = params["width"].as_u64() {
        args.push("-W".to_string());
        args.push(w.to_string());
    }
    if let Some(h) = params["height"].as_u64() {
        args.push("-H".to_string());
        args.push(h.to_string());
    }

    // Steps
    if let Some(steps) = params["steps"].as_u64() {
        args.push("--steps".to_string());
        args.push(steps.to_string());
    }

    // CFG scale
    if let Some(cfg) = params["cfgScale"].as_f64() {
        args.push("--cfg-scale".to_string());
        args.push(cfg.to_string());
    }

    // Seed
    if let Some(seed) = params["seed"].as_i64() {
        args.push("-s".to_string());
        args.push(seed.to_string());
    }

    // Sampling method
    if let Some(method) = params["samplingMethod"].as_str() {
        if !method.is_empty() {
            args.push("--sampling-method".to_string());
            args.push(method.to_string());
        }
    }

    // Flow shift
    if let Some(flow) = params["flowShift"].as_f64() {
        args.push("--flow-shift".to_string());
        args.push(flow.to_string());
    }

    // High noise parameters
    if let Some(steps) = params["highNoiseSteps"].as_u64() {
        args.push("--high-noise-steps".to_string());
        args.push(steps.to_string());
    }
    if let Some(cfg) = params["highNoiseCfgScale"].as_f64() {
        args.push("--high-noise-cfg-scale".to_string());
        args.push(cfg.to_string());
    }
    if let Some(method) = params["highNoiseSamplingMethod"].as_str() {
        if !method.is_empty() {
            args.push("--high-noise-sampling-method".to_string());
            args.push(method.to_string());
        }
    }

    // Init image for i2v
    if let Some(init) = params["initImage"].as_str() {
        if !init.is_empty() {
            args.push("-i".to_string());
            args.push(init.to_string());
        }
    }

    // Boolean flags
    if params["verbose"].as_bool() == Some(true) {
        args.push("-v".to_string());
    }
    if params["offloadToCpu"].as_bool() == Some(true) {
        args.push("--offload-to-cpu".to_string());
    }
    if params["clipOnCpu"].as_bool() == Some(true) {
        args.push("--clip-on-cpu".to_string());
    }
    if params["vaeOnCpu"].as_bool() == Some(true) {
        args.push("--vae-on-cpu".to_string());
    }
    if params["vaeTiling"].as_bool() == Some(true) {
        args.push("--vae-tiling".to_string());
    }

    // Output
    args.push("-o".to_string());
    args.push(output_path.to_string_lossy().to_string());

    Ok(args)
}
